app/controllers/application_controller.rb

```diff
@@ -1,10 +1,110 @@
+    @user = User.create(params)
+    session[:user_id] = @user.id #creates session attribute :user_id and sets to @user.id until logout
+    redirect '/tweets'
+  end
+
+  get '/login' do
+    logged_in? ? (redirect '/tweets') : (erb :'/users/login')
+  end
+
+  post '/login' do
+    redirect '/login' if params[:username].empty? || params[:password].empty?
+    @user = User.find_by(username: params[:username]) #find and set user (if possible)
+    if @user.authenticate(params[:password]) #create session (log user in) if supplied credentials are valid
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+  end
+end
+
+  get '/logout' do #logs user out and redirects to login page, otherwise redirects to homepage.
+    if logged_in?
+      session.clear
+      redirect '/login'
+    else
+      redirect '/'
+    end
+  end
+
+  get '/tweets' do #sends user to login page if not logged in, else loads tweet page
+    redirect '/login' if !logged_in?
+    @user = current_user #provides erb access to user and user's tweets
+    @tweets = @user.tweets
+    erb :'/tweets/tweets'
+  end
+
+  get '/tweets/new' do #Presents user with new tweet form, or sends them away if not logged in
+    logged_in? ? (erb :'/tweets/create_tweet') : (redirect '/login')
+  end
+
+  post '/tweets/new' do #creates tweet from new tweet form post request unless it's blank.
+    unless params[:content].empty?
+      @tweet = Tweet.create(user_id: current_user.id, content: params[:content]) #creates new tweet using current_user object id and params
+      redirect '/tweets'
+    end
+  end
+
+  get "/users/:slug" do
+    @user = find_by_slug(params[:slug])
+    erb :"/tweets/show_tweet"
+  end
+
+  get "/tweets/:id" do #redirects to tweet edit URL, made necessary because I misunderstood that the spec didn't want edit/delete on the same page.
+    logged_in? ? (redirect "/tweets/#{params[:id]}/edit") : (redirect "/login") #login check/redirect login is just to satisfy spec at this juncture
+  end
+
+  get "/tweets/:id/edit" do #finds tweet and renders edit page if the user is logged in, otherwise sends packing
+    redirect '/login' unless logged_in?
+    @tweet = Tweet.find(params[:id])
+    erb :'/tweets/edit_tweet'
+  end
+
+  post "/tweets/:id" do #Handles POST from tweet edit form and catches no-content exception
+    @tweet = Tweet.find(params[:id])
+    redirect "/tweets/#{@tweet.id}/edit" if params[:content].empty? #reloads page instead of saving tweet if tweet blank
+    @tweet.update(content: params[:content]) #otherwise modifies and saves it as expected, then redirects to /tweets
+    redirect '/tweets'
+  end
+
+  delete '/tweets/:id/delete' do #deletes tweet if if belongs to user, otherwise sends them away.
+    @tweet = Tweet.find(params[:id])
+    if current_user.id == @tweet.user_id
+      @tweet.delete
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  helpers do
+    def current_user
```

Comment:
Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id]) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/)).